cdap-api/src/main/java/co/cask/cdap/api/lineage/Read.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.lineage;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Represents READ operation from data source into the collection of fields.
+ */
+public class Read extends Operation {
+  private final EndPoint endPoint;
+  private final Set<String> fields;
+
+  private Read(String name, String description, EndPoint endPoint, Set<String> fields) {
+    super(name, description);
+    this.endPoint = endPoint;
+    this.fields = Collections.unmodifiableSet(new HashSet<>(fields));
+  }
+
+  /**
+   * Get the EndPoint representing the data source for this READ operation
+   * @return the EndPoint
+   */
+  public EndPoint getEndPoint() {
+    return endPoint;
+  }
+
+  /**
+   * Get the set of fields created by this READ operation
+   * @return the set of fields
+   */
+  public Set<String> getFields() {
+    return fields;
+  }
+
+  /**
+   * Builder for Read operation
+   */
+  public static class Builder extends Operation.Builder {
```

Comment:
can you write this:

```
new Read.Builder("readFile")
  .setDescription("reads from a file")
  .from(EndPoint.of("HRData")
  .to("offset", "body")
  .build();
```

I believe you can't because setDescription() returns a Operation.Builder. There is a better way to define builders for a hierarchy of subclasses. See for example, here: https://stackoverflow.com/questions/17164375/subclassing-a-java-builder-class

Reply:
Removed Builder. However thanks for sharing the link.. good to know.